From e9facd828215fee2fa637418bcb81ef65918bed6 Mon Sep 17 00:00:00 2001
From: Ssandhyarani <sirasanagandla.sandhyarani@ltts.com>
Date: Fri, 19 May 2023 17:53:30 +0530
Subject: [PATCH] REFPLTB-2394 : old override syntax issue in Kirkstone build

Reason for change: Since for kirkstone _append syntax is changed to :append,so updated the python scripts corresponding to the syntax
Test Procedure: In the yocto 4 kirkstone workpace, bitbake should not complain the old override syntax
Risks: None

Signed-off-by: Ssandhyarani <sirasanagandla.sandhyarani@ltts.com>
Change-Id: I61734a70351459018f524c7d6de74da3e54ef034
---
 bitbake/lib/bb/cache.py      |    6 +-
 bitbake/lib/bb/data_smart.py |  114 ++++++++++++++++++++++--------------------
 bitbake/lib/bb/siggen.py     |    5 +-
 3 files changed, 66 insertions(+), 59 deletions(-)

diff --git a/bitbake/lib/bb/cache.py b/bitbake/lib/bb/cache.py
index 92e9a3c..5772542 100644
--- a/bitbake/lib/bb/cache.py
+++ b/bitbake/lib/bb/cache.py
@@ -12,6 +12,8 @@
 # Copyright (C) 2003 - 2005 Michael 'Mickey' Lauer
 # Copyright (C) 2005        Holger Hans Peter Freyther
 # Copyright (C) 2005        ROAD GmbH
+# May 2023    sirasanagandla.sandhyarani@ltts.com(from arjun_daasuramdass@comcast.com) "_" should be allowed to override a variable.
+
 #
 # SPDX-License-Identifier: GPL-2.0-only
 #
@@ -54,12 +56,12 @@
 
     @classmethod
     def pkgvar(cls, var, packages, metadata):
-        return dict((pkg, cls.depvar("%s:%s" % (var, pkg), metadata))
+        return dict((pkg, cls.depvar("%s_%s" % (var, pkg), metadata))
                     for pkg in packages)
 
     @classmethod
     def taskvar(cls, var, tasks, metadata):
-        return dict((task, cls.getvar("%s:task-%s" % (var, task), metadata))
+        return dict((task, cls.getvar("%s_task-%s" % (var, task), metadata))
                     for task in tasks)
 
     @classmethod
diff --git a/bitbake/lib/bb/data_smart.py b/bitbake/lib/bb/data_smart.py
index dd20ca5..6d81db4 100644
--- a/bitbake/lib/bb/data_smart.py
+++ b/bitbake/lib/bb/data_smart.py
@@ -11,6 +11,7 @@
 # Copyright (C) 2005, 2006  Holger Hans Peter Freyther
 # Copyright (C) 2005        Uli Luckas
 # Copyright (C) 2005        ROAD GmbH
+# May 2023    sirasanagandla.sandhyarani@ltts.com(from arjun_daasuramdass@comcast.com) "_" should be allowed to override a variable.
 #
 # SPDX-License-Identifier: GPL-2.0-only
 #
@@ -26,8 +27,8 @@
 
 logger = logging.getLogger("BitBake.Data")
 
-__setvar_keyword__ = [":append", ":prepend", ":remove"]
-__setvar_regexp__ = re.compile(r'(?P<base>.*?)(?P<keyword>:append|:prepend|:remove)(:(?P<add>[^A-Z]*))?$')
+__setvar_keyword__ = ["_append", "_prepend", "_remove"]
+__setvar_regexp__ = re.compile(r'(?P<base>.*?)(?P<keyword>_append|_prepend|_remove)(_(?P<add>[^A-Z]*))?$')
 __expand_var_regexp__ = re.compile(r"\${[a-zA-Z0-9\-_+./~:]+?}")
 __expand_python_regexp__ = re.compile(r"\${@.+?}")
 __whitespace_split__ = re.compile(r'(\s)')
@@ -299,7 +300,7 @@
             for (r, override) in d.overridedata[var]:
                 for event in self.variable(r):
                     loginfo = event.copy()
-                    if 'flag' in loginfo and not loginfo['flag'].startswith(("_", ":")):
+                    if 'flag' in loginfo and not loginfo['flag'].startswith("_"):
                         continue
                     loginfo['variable'] = var
                     loginfo['op'] = 'override[%s]:%s' % (override, loginfo['op'])
@@ -374,7 +375,7 @@
         for event in history:
             if 'flag' in event:
                 continue
-            if event['op'] == ':remove':
+            if event['op'] == '_remove':
                 continue
             if isset and event['op'] == 'set?':
                 continue
@@ -537,15 +538,7 @@
 
     def setVar(self, var, value, **loginfo):
         #print("var=" + str(var) + "  val=" + str(value))
-
-        if not var.startswith("__anon_") and ("_append" in var or "_prepend" in var or "_remove" in var):
-            info = "%s" % var
-            if "file" in loginfo:
-                info += " file: %s" % loginfo["file"]
-            if "line" in loginfo:
-                info += " line: %s" % loginfo["line"]
-            bb.fatal("Variable %s contains an operation using the old override syntax. Please convert this layer/metadata before attempting to use with a newer bitbake." % info)
-
+        var = var.replace(":", "_")
         shortvar = var.split(":", 1)[0]
         if shortvar in self._var_renames:
             _print_rename_error(shortvar, loginfo, self._var_renames, fullvar=var)
@@ -580,7 +573,7 @@
             # pay the cookie monster
 
             # more cookies for the cookie monster
-            if ':' in var:
+            if '_' in var:
                 self._setvar_update_overrides(base, **loginfo)
 
             if base in self.overridevars:
@@ -591,27 +584,27 @@
             self._makeShadowCopy(var)
 
         if not parsing:
-            if ":append" in self.dict[var]:
-                del self.dict[var][":append"]
-            if ":prepend" in self.dict[var]:
-                del self.dict[var][":prepend"]
-            if ":remove" in self.dict[var]:
-                del self.dict[var][":remove"]
+            if "_append" in self.dict[var]:
+                del self.dict[var]["_append"]
+            if "_prepend" in self.dict[var]:
+                del self.dict[var]["_prepend"]
+            if "_remove" in self.dict[var]:
+                del self.dict[var]["_remove"]
             if var in self.overridedata:
                 active = []
                 self.need_overrides()
                 for (r, o) in self.overridedata[var]:
                     if o in self.overridesset:
                         active.append(r)
-                    elif ":" in o:
-                        if set(o.split(":")).issubset(self.overridesset):
+                    elif "_" in o:
+                        if set(o.split("_")).issubset(self.overridesset):
                             active.append(r)
                 for a in active:
                     self.delVar(a)
                 del self.overridedata[var]
 
         # more cookies for the cookie monster
-        if ':' in var:
+        if '_' in var:
             self._setvar_update_overrides(var, **loginfo)
 
         # setting var
@@ -637,8 +630,8 @@
 
     def _setvar_update_overrides(self, var, **loginfo):
         # aka pay the cookie monster
-        override = var[var.rfind(':')+1:]
-        shortvar = var[:var.rfind(':')]
+        override = var[var.rfind('_')+1:]
+        shortvar = var[:var.rfind('_')]
         while override and __override_regexp__.match(override):
             if shortvar not in self.overridedata:
                 self.overridedata[shortvar] = []
@@ -647,9 +640,9 @@
                 self.overridedata[shortvar] = list(self.overridedata[shortvar])
                 self.overridedata[shortvar].append([var, override])
             override = None
-            if ":" in shortvar:
-                override = var[shortvar.rfind(':')+1:]
-                shortvar = var[:shortvar.rfind(':')]
+            if "_" in shortvar:
+                override = var[shortvar.rfind('_')+1:]
+                shortvar = var[:shortvar.rfind('_')]
                 if len(shortvar) == 0:
                     override = None
 
@@ -660,6 +653,8 @@
         """
         Rename the variable key to newkey
         """
+        key = key.replace(":", "_")
+        newkey = newkey.replace(":", "_")
         if key == newkey:
             bb.warn("Calling renameVar with equivalent keys (%s) is invalid" % key)
             return
@@ -689,7 +684,7 @@
                 self.overridedata[newkey].append([v.replace(key, newkey), o])
                 self.renameVar(v, v.replace(key, newkey))
 
-        if ':' in newkey and val is None:
+        if '_' in newkey and val is None:
             self._setvar_update_overrides(newkey, **loginfo)
 
         loginfo['variable'] = key
@@ -701,14 +696,15 @@
     def appendVar(self, var, value, **loginfo):
         loginfo['op'] = 'append'
         self.varhistory.record(**loginfo)
-        self.setVar(var + ":append", value, ignore=True, parsing=True)
+        self.setVar(var + "_append", value, ignore=True, parsing=True)
 
     def prependVar(self, var, value, **loginfo):
         loginfo['op'] = 'prepend'
         self.varhistory.record(**loginfo)
-        self.setVar(var + ":prepend", value, ignore=True, parsing=True)
+        self.setVar(var + "_prepend", value, ignore=True, parsing=True)
 
     def delVar(self, var, **loginfo):
+        var = var.replace(":", "_")
         self.expand_cache = {}
 
         loginfo['detail'] = ""
@@ -717,9 +713,9 @@
         self.dict[var] = {}
         if var in self.overridedata:
             del self.overridedata[var]
-        if ':' in var:
-            override = var[var.rfind(':')+1:]
-            shortvar = var[:var.rfind(':')]
+        if '_' in var:
+            override = var[var.rfind('_')+1:]
+            shortvar = var[:var.rfind('_')]
             while override and override.islower():
                 try:
                     if shortvar in self.overridedata:
@@ -729,13 +725,14 @@
                 except ValueError as e:
                     pass
                 override = None
-                if ":" in shortvar:
-                    override = var[shortvar.rfind(':')+1:]
-                    shortvar = var[:shortvar.rfind(':')]
+                if "_" in shortvar:
+                    override = var[shortvar.rfind('_')+1:]
+                    shortvar = var[:shortvar.rfind('_')]
                     if len(shortvar) == 0:
                          override = None
 
     def setVarFlag(self, var, flag, value, **loginfo):
+        var = var.replace(":", "_")
         self.expand_cache = {}
 
         if var == "BB_RENAMED_VARIABLES":
@@ -754,7 +751,7 @@
             self._makeShadowCopy(var)
         self.dict[var][flag] = value
 
-        if flag == "_defaultval" and ':' in var:
+        if flag == "_defaultval" and '_' in var:
             self._setvar_update_overrides(var, **loginfo)
         if flag == "_defaultval" and var in self.overridevars:
             self._setvar_update_overridevars(var, value)
@@ -767,6 +764,7 @@
             self.dict["__exportlist"]["_content"].add(var)
 
     def getVarFlag(self, var, flag, expand=True, noweakdefault=False, parsing=False, retparser=False):
+        var = var.replace(":", "_")
         if flag == "_content":
             cachename = var
         else:
@@ -786,11 +784,11 @@
             active = {}
             self.need_overrides()
             for (r, o) in overridedata:
-                # FIXME What about double overrides both with "_" in the name?
+                # What about double overrides both with "_" in the name?
                 if o in self.overridesset:
                     active[o] = r
-                elif ":" in o:
-                    if set(o.split(":")).issubset(self.overridesset):
+                elif "_" in o:
+                    if set(o.split("_")).issubset(self.overridesset):
                         active[o] = r
 
             mod = True
@@ -798,10 +796,10 @@
                 mod = False
                 for o in self.overrides:
                     for a in active.copy():
-                        if a.endswith(":" + o):
+                        if a.endswith("_" + o):
                             t = active[a]
                             del active[a]
-                            active[a.replace(":" + o, "")] = t
+                            active[a.replace("_" + o, "")] = t
                             mod = True
                         elif a == o:
                             match = active[a]
@@ -820,12 +818,12 @@
                 value = copy.copy(local_var["_defaultval"])
 
 
-        if flag == "_content" and local_var is not None and ":append" in local_var and not parsing:
+        if flag == "_content" and local_var is not None and "_append" in local_var and not parsing:
             self.need_overrides()
-            for (r, o) in local_var[":append"]:
+            for (r, o) in local_var["_append"]:
                 match = True
                 if o:
-                    for o2 in o.split(":"):
+                    for o2 in o.split("_"):
                         if not o2 in self.overrides:
                             match = False                            
                 if match:
@@ -833,13 +831,13 @@
                         value = ""
                     value = value + r
 
-        if flag == "_content" and local_var is not None and ":prepend" in local_var and not parsing:
+        if flag == "_content" and local_var is not None and "_prepend" in local_var and not parsing:
             self.need_overrides()
-            for (r, o) in local_var[":prepend"]:
+            for (r, o) in local_var["_prepend"]:
 
                 match = True
                 if o:
-                    for o2 in o.split(":"):
+                    for o2 in o.split("_"):
                         if not o2 in self.overrides:
                             match = False                            
                 if match:
@@ -853,12 +851,12 @@
         if expand:
             value = parser.value
 
-        if value and flag == "_content" and local_var is not None and ":remove" in local_var and not parsing:
+        if value and flag == "_content" and local_var is not None and "_remove" in local_var and not parsing:
             self.need_overrides()
-            for (r, o) in local_var[":remove"]:
+            for (r, o) in local_var["_remove"]:
                 match = True
                 if o:
-                    for o2 in o.split(":"):
+                    for o2 in o.split("_"):
                         if not o2 in self.overrides:
                             match = False                            
                 if match:
@@ -894,6 +892,7 @@
         return value
 
     def delVarFlag(self, var, flag, **loginfo):
+        var = var.replace(":", "_")
         self.expand_cache = {}
 
         local_var, _ = self._findVar(var)
@@ -911,6 +910,7 @@
             del self.dict[var][flag]
 
     def appendVarFlag(self, var, flag, value, **loginfo):
+        var = var.replace(":", "_")
         loginfo['op'] = 'append'
         loginfo['flag'] = flag
         self.varhistory.record(**loginfo)
@@ -918,6 +918,7 @@
         self.setVarFlag(var, flag, newvalue, ignore=True)
 
     def prependVarFlag(self, var, flag, value, **loginfo):
+        var = var.replace(":", "_")
         loginfo['op'] = 'prepend'
         loginfo['flag'] = flag
         self.varhistory.record(**loginfo)
@@ -925,6 +926,7 @@
         self.setVarFlag(var, flag, newvalue, ignore=True)
 
     def setVarFlags(self, var, flags, **loginfo):
+        var = var.replace(":", "_")
         self.expand_cache = {}
         infer_caller_details(loginfo)
         if not var in self.dict:
@@ -939,12 +941,13 @@
             self.dict[var][i] = flags[i]
 
     def getVarFlags(self, var, expand = False, internalflags=False):
+        var = var.replace(":", "_")
         local_var, _ = self._findVar(var)
         flags = {}
 
         if local_var:
             for i in local_var:
-                if i.startswith(("_", ":")) and not internalflags:
+                if i.startswith("_") and not internalflags:
                     continue
                 flags[i] = local_var[i]
                 if expand and i in expand:
@@ -955,6 +958,7 @@
 
 
     def delVarFlags(self, var, **loginfo):
+        var = var.replace(":", "_")
         self.expand_cache = {}
         if not var in self.dict:
             self._makeShadowCopy(var)
@@ -1043,8 +1047,8 @@
             for (r, o) in self.overridedata[var]:
                 if o in self.overridesset:
                     overrides.add(var)
-                elif ":" in o:
-                    if set(o.split(":")).issubset(self.overridesset):
+                elif "_" in o:
+                    if set(o.split("_")).issubset(self.overridesset):
                         overrides.add(var)
 
         for k in keylist(self.dict):
diff --git a/bitbake/lib/bb/siggen.py b/bitbake/lib/bb/siggen.py
index 0a9ce0e..aa8180b 100644
--- a/bitbake/lib/bb/siggen.py
+++ b/bitbake/lib/bb/siggen.py
@@ -1,5 +1,6 @@
 #
 # Copyright BitBake Contributors
+# May 2023    sirasanagandla.sandhyarani@ltts.com(from arjun_daasuramdass@comcast.com) "_" should be allowed to override a variable.
 #
 # SPDX-License-Identifier: GPL-2.0-only
 #
@@ -245,7 +246,7 @@
         #    self.dump_sigtask(fn, task, d.getVar("STAMP"), False)
 
         for task in taskdeps:
-            d.setVar("BB_BASEHASH:task-%s" % task, self.basehash[fn + ":" + task])
+            d.setVar("BB_BASEHASH_task-%s" % task, self.basehash[fn + ":" + task])
 
     def postparsing_clean_cache(self):
         #
@@ -345,7 +346,7 @@
 
         h = hashlib.sha256(data.encode("utf-8")).hexdigest()
         self.taskhash[tid] = h
-        #d.setVar("BB_TASKHASH:task-%s" % task, taskhash[task])
+        #d.setVar("BB_TASKHASH_task-%s" % task, taskhash[task])
         return h
 
     def writeout_file_checksum_cache(self):
